### DISCUSSION TITLE
Log4j 2 module deprecation

### DISCUSSION THREAD

**Discussion by vy** [apache/logging-log4j2#1822] [Ideas]
_[This message is cross-post of [a mailing list thread](https://lists.apache.org/thread/76f260b933ol7og757123lyl1ckjvm8y). Depending on your preference, you can participate either here or there.]_

We always strive to only release industrial-strength components.

Following an internal discussion and looking at statistics from several sources, the PMC plans to deprecate in the `2.x` version several components for removal in `3.x`.

We know that statistics are not everything and that is why :loudspeaker: **we would like to hear your opinion on these modules**. The discussion will be open for a month, after which the PMC will announce a definitive list of deprecations.

## Cassandra appender (`log4j-cassandra`)

We support Cassandra [3.x, which will reach end-of-life when 4.2.x is released](https://cassandra.apache.org/_/blog/Behind-the-scenes-of-an-Apache-Cassandra-Release.html). Since the stats show marginal usage of the module, we expect neither a `log4j-cassandra4`.

## CouchDB appender (`log4j-couchdb`)

It is even less used than Cassandra according to the stats.

## Docker lookup (`log4j-docker`)

Seldom used and standard practice is generally to log to the console and let the Docker environment add the data.

## GELF layout (`GelfLayout` in `log4j-core`)

It is mostly superseded by the JSON Template Layout.

## Kafka appender (`KafkaAppender` in `log4j-core`)

It contains several critical bug reports neither the PMC, nor the community is willing to deal with.

## Kubernetes lookup (`log4j-kubernetes`)

Seldom used and standard practice is generally to log to the
console and let the Docker environment add the data.

## ZeroMQ appender (`JeroMqAppender` in `log4j-core`)

Since version 2.6 it had only one PR/bug report and it was from PAX Logging maintainer. There is [an alternative from a ZeroMQ maintainer, Fabrice Bacchella](https://github.com/fbacchella/loghublog4j2), that provides important features like encryption or choice of socket type.

## JNDI-related features

JNDI is an old technology from the times of ponies and unicorns, when nobody cared about security. The Error Prone team decided to mark [JNDI usage as error by default](https://errorprone.info/bugpattern/BanJNDI).

## JPA appender (`log4j-jpa`)

Has marginal downloads and requires a migration to Jakarta EE. We think that `log4j-jdbc` (which is also seldom used) provides a good alternative for SQL databases.

## Jackson-based layouts (`Json-`, `Xml-`, and `YamlLayout` in `log4j-core`)

`JsonTemplateLayout` supersedes `JsonLayout`; `XmlLayout` is seldom used, and `YamlLayout`, as far as we are concerned, is not be used at all.

## MongoDB 3 appender (`log4j-mongodb3`)

Support for MongoDB 3 ended two years ago.

## `log4j-spring-boot`

Its features are included in Spring Boot `3.x`.

## SMTP appender (`SmtpAppender` in `log4j-core`)

Its usage is at the bottom of the stats.

## `log4j-taglib`

Needs a Jakarta version, but it is based on a legacy technology that is currently seldom used.

**Comment by garydgregory** [comment]
This all looks OK to me. The nice-to-have would be a migration guide.

**Comment by fbacchella** [comment]
If any one is interested in features provided by the JeroMqAppender but that are still missing in my own ZMQ appender, please fill free to contact me.

**Comment by vy** [comment]
[The resolution is being voted.](https://lists.apache.org/thread/szvpxkgpvc1kgz4zbm4m9crwv3ml9xpg)
